### PR TITLE
Added OSD and Host replacement scenarios in stretch mode

### PR DESCRIPTION
**Pull request by pdhiran**
# Description

Polarion : https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83575474 

Automation of Host and OSD addition/ Removal in stretch mode.

Tests included:
1. Removal of 1 OSD from data site and check the warning generated for imbalanced weights.
2. Addition of 1 OSD from data site and check the warning generated for imbalanced weights is removed.
3. Removal of 1 OSD Host from data site and check the warning generated for imbalanced weights.
2. Addition of 1 OSD Host from data site and check the warning generated for imbalanced weights is removed.

### CODE REVIEW

File: suites/quincy/rados/tier-3_rados_test-location-stretch-mode.yaml [hunk 1]
```diff
@@ -317,16 +317,28 @@ tests:
...
-#      name: Netsplit Scenarios data-data sites
-#      module: test_stretch_netsplit_scenarios.py
-#      polarion-id: CEPH-83574979
-#      config:
-#        pool_name: test_stretch_pool8
-#        netsplit_site: DC1
-#        tiebreaker_mon_site_name: arbiter
-#        delete_pool: true
-#      desc: Test stretch Cluster netsplit scenario between data sites
+# New Bug: https://bugzilla.redhat.com/show_bug.cgi?id=2265116
```
Comment (pdhiran) [inline suites/quincy/rados/tier-3_rados_test-location-stretch-mode.yaml:320]: add bz in the failure log.

File: tests/rados/test_stretch_osd_serviceability_scenarios.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    method_should_succeed(wait_for_device, test_host, target_osd, action="add")
+    time.sleep(30)
+    log.debug(
+        "Completed addition of OSD post removal. Checking for inactive PGs post OSD addition"
+    )
+    if not rados_obj.check_inactive_pgs_on_pool(pool_name=pool_name):
+        log.error(f"Inactive PGs found on pool : {pool_name}")
+        inactive_pgs += 1
+
+    if inactive_pgs > 10:
```
Comment (SrinivasaBharath) [inline tests/rados/test_stretch_osd_serviceability_scenarios.py:199]: Any reason for picking the inactive PG count to 10?

File: tests/rados/test_stretch_osd_serviceability_scenarios.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        log.error("Found inactive PGs on the cluster during OSD removal/Addition")
+        raise Exception("Inactive PGs during stop error")
+
+    # Checking if the expected health warning about the different site weights is removed post addition of OSD.
+    if check_stretch_health_warning():
+        log.error("Warnings not removed on the cluster post osd addition")
+        raise Exception("Warning present on cluster")
+
+    # Checking cluster health after OSD removal
```
Comment (neha-gangadhar) [inline tests/rados/test_stretch_osd_serviceability_scenarios.py:208]: minor comment in section - checking cluster health after OSD addition.

File: tests/rados/test_stretch_osd_serviceability_scenarios.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            crush_bucket_val=test_host_site,
+        )
+    except Exception as err:
+        log.error(
+            f"Could not add host : {test_host.hostname} into the cluster and deploy OSDs. Error : {err}"
+        )
+        raise Exception("Host not added error")
+    time.sleep(60)
+
+    # Waiting for recovery to post OSD host removal
```
Comment (neha-gangadhar) [inline tests/rados/test_stretch_osd_serviceability_scenarios.py:288]: Minor comment in section - Waiting for recovery to post OSD host addition

File: tests/rados/test_stretch_osd_serviceability_scenarios.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    if not rados_obj.check_inactive_pgs_on_pool(pool_name=pool_name):
+        log.error(f"Inactive PGs found on pool : {pool_name}")
+        inactive_pgs += 1
+    method_should_succeed(wait_for_clean_pg_sets, rados_obj, timeout=12000)
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    if inactive_pgs > 10:
+        log.error("Found inactive PGs on the cluster during OSD removal/Addition")
+        raise Exception("Inactive PGs during stop error")
+
+    # Checking if the expected health warning about the different site weights is displayed post removal of OSD host
```
Comment (neha-gangadhar) [inline tests/rados/test_stretch_osd_serviceability_scenarios.py:298]: please modify to post addition of OSD host

**Review by neha-gangadhar — COMMENTED** [review]
Minor comments in the "comments section" of the code. please review and fix them. good to merge.

File: tests/rados/test_stretch_osd_serviceability_scenarios.py (hunk 2 above)
Comment (pdhiran) [reply tests/rados/test_stretch_osd_serviceability_scenarios.py:199]: Just added it as failsafe number, that if we have more than that many inactive PGs at any point in a pool. The Pgs state is checked at that moment, and not overtime to see if a PG is in inactive state for long durations.

File: tests/rados/test_stretch_osd_serviceability_scenarios.py (hunk 3 above)
Comment (pdhiran) [reply tests/rados/test_stretch_osd_serviceability_scenarios.py:208]: Updated.

File: tests/rados/test_stretch_osd_serviceability_scenarios.py (hunk 4 above)
Comment (pdhiran) [reply tests/rados/test_stretch_osd_serviceability_scenarios.py:288]: Updated the comment.

File: tests/rados/test_stretch_osd_serviceability_scenarios.py (hunk 5 above)
Comment (pdhiran) [reply tests/rados/test_stretch_osd_serviceability_scenarios.py:298]: Updated the comment.

**Comment by pdhiran** [comment]
Updated all review comments.

**Review by neha-gangadhar — APPROVED** [review]
approved

**Pull request merged**